pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/WindowFunction.java

```diff
@@ -0,0 +1,34 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.functions.api;
+
+import java.util.Collection;
+
+/**
+ * This is the interface of the windowed function api. The process method is called
+ * for every triggered window.
+ */
+@FunctionalInterface
+public interface WindowFunction<I, O> {
+    /**
+     * Process the input.
+     * @return the output
+     */
+    O process(Collection<I> input, WindowContext context) throws Exception;
```

Comment:
I don't know if this is the API we should have for window functions because there is no way to get the topic of individual msg/record in the collection/window.  Maybe the collection should be something like Collection<Record<I>> ? so that users can get the topic from which the msg originated from.

Reply:
In a follow up pr, I was thinking of adding something to WindowContext to get something like getCurrentWindow(or something like that) that returns information in terms of Collection<Record>

Reply:
but the thing is each entry in the collection could be from a different topic.  I am not sure how a getCurrentWindow function would work.

Reply:
If getcurrentWindow returns Collection<Record> that should be sufficient. Each Record will have information about the topic of that record.

Reply:
but collections are not ordered in Java

Reply:
that is also not a eloquent API IMO.  Users would have to use the index from one list to look up info in another list manually

Reply:
So the current java.util.function way for handling windowing is not ordered in Java either.
WRT API, if the WindowContext exposes a getCurrentWindow that exposes a Window<Record> that should deal with most of the use cases. For the most simple of use cases, just use the collection<I>. For more complicated uses, rely instead on windowContext.getCurrentWindow()

Reply:
@jerrypeng as discussed, I've updated the interface to reflect our discussions. Please take a look again. Thanks!